packages/rest/src/http-handler.ts

```diff
@@ -33,8 +34,12 @@ export class HttpHandler {
     this.handleRequest = (req, res) => this._handleRequest(req, res);
   }
 
-  registerController(name: ControllerClass, spec: ControllerSpec) {
-    this._routes.registerController(name, spec);
+  registerController(
+    controllerCtor: ControllerClass,
+    spec: ControllerSpec,
+    factory?: ControllerFactory,
```

Comment:
This API is too loose to my taste. For example, it's possible to register a controller in such way that  `factory` is returning an instance of a different class than `controllerCtor`:

```ts
registerController(
  ProductController,
  spec,
  ctx => new CategoryController());
```

IIUC, the only reason you need a direct access to `controllerCtor` is to obtain the controller name if the spec does not provide it. If that's correct, then I am proposing to modify this API as follows:

 1. Always require the controller factory.
 2. Replace required `controllerCtor` with an optional controller name argument.

```ts
registerController(
  factory: ControllerFactory,
  spec: ControllerSpec,
  name?: string,
)
```

A similar change should be applied to `ControllerRoute` class.

Thoughts?

Reply:
Controller constructor is still critical to look up corresponding metadata for some of the actions. That's why we bind it to `controller.current.ctor` at the moment.

Reply:
We typically decorate controller classes and their members to provider metadata for actions such as `authenticate`. Being able to access the controller constructor gives us access to metadata associated with the resolved controller class.

Reply:
Did you see https://github.com/strongloop/loopback-next/pull/993/commits/2adcaff167fe7edf5c6aa43cf28187e99fe4347e?